cpp/include/raft/core/device_coo_matrix.hpp

```diff
@@ -0,0 +1,236 @@
+                                                         true,
+                                                         ContainerPolicy,
+                                                         SparsityType::PRESERVING>;
+
+/**
+ * Specialization for a sparsity-owning coordinate structure which uses device memory
+ */
+template <typename RowType,
+          typename ColType,
+          typename NZType,
+          template <typename T> typename ContainerPolicy = detail::device_uvector_policy>
+using device_coordinate_structure =
+  coordinate_structure<RowType, ColType, NZType, true, ContainerPolicy>;
+
+/**
+ * Specialization for a sparsity-preserving coordinate structure view which uses device memory
+ */
+template <typename RowType, typename ColType, typename NZType>
+using device_coordinate_structure_view = coordinate_structure_view<RowType, ColType, NZType, true>;
+
+/**
+ * Create a sparsity-owning sparse matrix in the coordinate format. sparsity-owning means that
+ * all of the underlying vectors (data, indptr, indices) are owned by the coo_matrix instance. If
+ * not known up front, the sparsity can be ignored in this factory function and `resize()` invoked
+ * on the instance once the sparsity is known.
+ * @tparam ElementType
+ * @tparam RowType
+ * @tparam ColType
+ * @tparam NZType
+ * @param[in] handle a raft handle for managing expensive device resources
+ * @param[in] n_rows total number of rows in the matrix
+ * @param[in] n_cols total number of columns in the matrix
+ * @param[in] nnz number of non-zeros in the matrix if known [optional]
+ * @return a sparsity-owning sparse matrix in coordinate (coo) format
+ */
+template <typename ElementType, typename RowType, typename ColType, typename NZType>
+auto make_device_coo_matrix(raft::device_resources const& handle,
+                            RowType n_rows,
+                            ColType n_cols,
+                            NZType nnz = 0)
+{
+  return device_sparsity_owning_coo_matrix<ElementType, RowType, ColType, NZType>(
+    handle, n_rows, n_cols, nnz);
+}
+
+/**
+ * Create a sparsity-preserving sparse matrix in the coordinate format. sparsity-preserving means
+ * that a view of the coo sparsity is supplied, allowing the values in the sparsity to change but
+ * not the sparsity itself. The csr_matrix instance does not own the sparsity, the sparsity must
+ * be known up front, and cannot be resized later.
+ * @tparam ElementType
+ * @tparam RowType
+ * @tparam ColType
+ * @tparam NZType
+ * @param[in] handle raft handle for managing expensive device resources
+ * @param[in] structure_ a sparsity-preserving coordinate structural view
+ * @return a sparsity-preserving sparse matrix in coordinate (coo) format
+ */
+template <typename ElementType, typename RowType, typename ColType, typename NZType>
+auto make_device_coo_matrix(raft::device_resources const& handle,
+                            device_coordinate_structure_view<RowType, ColType, NZType> structure_)
```

Comment:
Should this already be a `shared_ptr`, and we produce views we return a `shared_ptr` to users?

Reply:
No, the views are trivially copyable by design so we don't want to have to burden the user to have to produce a `shared_ptr` here. The `shared_ptr` is more of an implementation detail from the perspective of the `coo_matrix` because it's really only being used to unify the preserving and owning variants underneath without having to introduce bloat.